File: src/main/java/com/github/thorben/webnd/foundation/AbstractEntity.java
package com.github.thorben.webnd.foundation;

import javax.persistence.MappedSuperclass;
import java.io.Serializable;

@MappedSuperclass
public abstract class AbstractEntity<T> implements Serializable {

	public abstract T getId();

	@Override
	public int hashCode() {
		T id = getId();
		return id == null ? super.hashCode() : id.hashCode();
	}

	@Override
	public boolean equals(Object that) {
		if (this == that) {
			return true;
		}

		if (!(that instanceof AbstractEntity) ||
				!(that.getClass().isAssignableFrom(getClass())) ||
				getClass().isAssignableFrom(that.getClass())) {
			return false;
		}

		AbstractEntity<T> valueObject = (AbstractEntity<T>) that;
		return getId() != null && valueObject.getId().equals(getId());
	}

	@Override
	public String toString() {
		return getClass().getSimpleName() + "#" + getId();
	}

}

File: src/main/kotlin/com/github/thorben/webnd/foundation/AbstractEntity.kt
package com.github.thorben.webnd.foundation

import java.io.Serializable
import javax.persistence.MappedSuperclass

@MappedSuperclass
abstract class AbstractEntity<T> : Serializable {
	abstract val id: T
	override fun hashCode(): Int {
		val id = id
		return id?.hashCode() ?: super.hashCode()
	}

	override fun equals(other: Any?): Boolean {
		if (this === other) {
			return true
		}
		if (other !is AbstractEntity<*> ||
				!other.javaClass.isAssignableFrom(javaClass) ||
				javaClass.isAssignableFrom(other.javaClass)) {
			return false
		}
		val valueObject = other as AbstractEntity<T>
		return id != null && valueObject.id == id
	}

	override fun toString(): String {
		return javaClass.simpleName + "#" + id
	}
}